build as a bin and lib.

trgill opened the pull request:
Move main.rs to bin/stratisd.rs to allow for building stratisd as bin.

Added lib.rs so that the remainder of the application is compiled as a
lib crate.

Added two sample integration tests blockdev_tests.rs and pool_tests.rs.

Added [dev-dependencies] to include Json parser, log, and env_logger for
use in the integration tests.

Note: see README.md in the tests directory for instructions on enabling test logging.

Log levels currently supported are debug and info.

Added test_config.json to hold parameters for tests.  Currently it only
has 1 array element "ok_to_destroy_dev_array_key".  This array contains a
a list of devices that can be safely destroyed during testing.

Moved dbgp! macro to stratisd.rs - it is the only place it is used - we
should probably switch to a logging crate.

Signed-off-by: Todd Gill <tgill@redhat.com>

trgill commented:
@mulkieran @agrover review please?

FYI - I had to put a #[ignore] tag on the test_pools().  We can remove when we implement cleanup of the disk label for destroy_pool().

trgill commented on Cargo.toml:
```diff
@@ -29,3 +29,6 @@ features = ["v4"]
 quickcheck = "0.4"
+rustc-serialize = "0.3.21"
+log = "0.3"
+env_logger="0.3.5"
```
Will add newline here.

mulkieran replied:
Is your comment above totally obsolete?

agrover commented:
Wow you've been busy! 😃

- I think some of the stuff about running the tests, log levels etc deserves to have greater visibility than the checkin comment. Put it in the README.md?
- This one commit is pretty big. Are there ways it could separate logical changes into separate commits more? Moving dbgp, adding deps and updating Cargo.lock, making into a lib, test framework, the initial tests themselves -- these all build on each other but seem logically distinct to me.
- in the commitlog for the libification, can you take a stab at discussing the other methods you investigated before settling on this approach? Having a record of why this is better than the Cargo approach will be great if we get questions from contributors down the line.

Much thanks! This is going to be so nice to have this in place.

trgill commented:
@agrover  Ok - thanks.  Will do.

trgill commented:
@agrover I broke up the commit into steps that would compile and follow your suggestions as much as possible.

mulkieran commented:
@trgill Is there some preliminary code for releasing the pools properly in the works?

trgill commented:
@mulkieran I don't have any code for releasing the pools properly.  I talked with @agrover about the issue briefly.  I think he is considering some alternatives based on how recoverable we want to make deleted pools.

trgill commented on tests/blockdev_tests.rs:
```diff
@@ -0,0 +1,89 @@
+        let dev = try!(Device::from_str(&path.to_string_lossy()));
+        devices.insert(dev);
+        devices_copy.insert(dev.clone());
+    }
```
I need to change this to use blockdev::resolve_devices().  Will do that now.

mulkieran commented on src/lib.rs:
```diff
@@ -0,0 +1,34 @@
+pub mod stratis;
+pub mod engine;
+pub mod dbus_api;
+#[allow(dead_code)]
```
What dead code does this allow? Should the dead code get removed?

trgill replied:
Fixed

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -61,6 +52,17 @@ fn write_err(err: StratisError) -> StratisResult<()> {
     Ok(())
 }
 
+pub static mut DEBUG: bool = false;
```
If ```dbgp!``` was moved, rather than copied, shouldn't there be a corresponding ```-``` entry in this commit?

mulkieran replied:
Ok, looks good now.

mulkieran commented on tests/util/test_result.rs:
```diff
@@ -0,0 +1,82 @@
+
+#[derive(Debug)]
+pub enum TestErrorEnum {
+    _Ok,
```
Might as well get rid of constructors with underscores. Also, it is possible that ```_Ok``` will always be unnecessary, for Ok is by definition not an error.

mulkieran commented on tests/util/test_result.rs:
```diff
@@ -0,0 +1,82 @@
+    _NotFound(String),
+}
+
+impl TestErrorEnum {
```
See https://github.com/stratis-storage/stratisd/pull/207/commits/b2ab75b804b7c3cdee9ce0c0107a2d44d8883df0 where some improvements are made to the existing error structure. Probably this commit should get off on the right foot by including those improvements.

mulkieran commented on tests/util/test_result.rs:
```diff
@@ -0,0 +1,82 @@
+    }
+}
+
+// Define a common error enum.
```
Please remove this comment.

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -2,7 +2,7 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-#![allow(dead_code)] // only temporary, until more stuff is filled in
+// only temporary, until more stuff is filled in
```
Well, rustfmt did us a favor here. You can drop this commit with PR #209, but now we know that you can still remove the ```#![allow(dead_code)]``` directive and accompanying comment in one of these commits.

mulkieran commented on tests/util/blockdev_utils.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+#[allow(dead_code)]
```
You should be able to drop this directive when you add the commit that uses this method.

mulkieran commented on tests/util/blockdev_utils.rs:
```diff
@@ -0,0 +1,87 @@
+    Ok(())
+}
+
+#[allow(dead_code)]
```
Looks like this method can be private from the start, so  dead code directive never needed.

mulkieran commented on tests/util/blockdev_utils.rs:
```diff
@@ -0,0 +1,87 @@
+    Ok(())
+}
+
+#[allow(dead_code)]
```
This function isn't used anywhere in the PR, can you remove it?

trgill commented:
@mulkieran I think I've addressed your comments except for 1.

If I don't tag blockdev_utils::get_ownership() with #[allow(dead_code)], I get a warning on the test compile.  The function is called in a test, I tried tagging with a

mulkieran commented on tests/util/blockdev_utils.rs:
```diff
@@ -0,0 +1,74 @@
+    Ok(ownership)
+}
+
+pub fn clean_blockdevs_headers(blockdev_paths: &Vec<&Path>) -> TestResult<()> {
```
Seems like this has too many plurals in the name, ```clean_blockdev_headers``` might seem more natural. Also, it might be possible to substitute for this method a map and then a find, or something.

mulkieran commented on tests/util/blockdev_utils.rs:
```diff
@@ -0,0 +1,74 @@
+}
+
+#[allow(dead_code)] // Needed to avoid warning on test build
+pub fn get_ownership(path: &PathBuf) -> TestResult<DevOwnership> {
```
I think the warning is caused because this method is only used in a test that is currently ignored. The comment could say that.

mulkieran commented on tests/util/test_config.rs:
```diff
@@ -0,0 +1,127 @@
+
+pub struct TestConfig {
+    filename: String,
+    json: Option<Json>,
```
Please use a name that defines the thing, not its format.

mulkieran commented on tests/util/test_result.rs:
```diff
@@ -0,0 +1,76 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+extern crate rustc_serialize;
```
Everything in this file makes sense to me.

mulkieran commented on tests/util/test_config.rs:
```diff
@@ -0,0 +1,127 @@
+        Ok(ok_to_destroy_dev_list)
+    }
+
+    pub fn read_test_config(&mut self) -> TestResult<()> {
```
A comment would be helpful here.

mulkieran replied:
I think you can make this method private.

mulkieran commented on tests/util/test_config.rs:
```diff
@@ -0,0 +1,127 @@
+    }
+
+
+    pub fn get_safe_to_destroy_devs(&mut self) -> TestResult<Vec<String>> {
```
Method header comment helpful here.

mulkieran commented on tests/util/test_config.rs:
```diff
@@ -0,0 +1,127 @@
+    json: Option<Json>,
+}
+
+macro_rules! test_try_get_destroy_devs {
```
Maybe this could be a function, not a macro, as well.

trgill commented:
RE:  removing OK_TO_DESTROY_DEV_KEY and using string literal

I think we are going to be adding other sections to the JSON.  It would be best to keep the keys in one place.

trgill commented:
How is PR #207 is going to make the tests obsolete?

One test is ignored because it fails due to a bug  releasing devices that needs to be fixed.  I'll remove the #ignore tag on it.  It won't run when the JSON is empty anyway.  But when we use it on real devices it will fail.

mulkieran commented:
I don't mean the tests will be obsolete, I mean that the code they rely on will be. So, tests will have to be fixed up immediately in order to compile by PR #207.

trgill commented:
I think the adjustments for PR #207 will be mostly to the code in blockdev_utils.rs.  I'd suggest we merge this first - then 207.

mulkieran commented:
Ok! :+1: .
